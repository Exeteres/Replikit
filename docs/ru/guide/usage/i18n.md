# Интернационализация

Пакет `@replikit/i18n` предоставляет простую строготипизированную систему интернационализации.

## Базовое использование

Для объявления ресурса надо расширить интерфейс `Locale`:

```ts
import "@replikit/i18n/typings";

declare module "@replikit/i18n/typings/locale" {
    export interface Locale {
        mymodule: {
            hello: string;
        }
    }
}
```

И добавить перевод для конкретного языка:

```ts
import { locales } from "@replikit/i18n";

locales.add("ru", "mymodule", {
    hello: "Привет!"
});
```

После чего можно использовать локализацию в контексте маршрутизатора:

```ts
import { router } from "@replikit/router";

router.of("message:received").use(context => {
    context.reply(context.t.mymodule.hello);
});
```

## Форматирование

Поскольку локализация осуществляется обычными JS объектами, помимо строк можно также использовать функции.
Вместе с синтаксисом ES6 шаблонов получается полноценное форматирование:

```ts
export interface Locale {
    mymodule: {
        hello(name: string): string;
    }
}

locales.add("ru", "mymodule", {
    hello: name => `Привет, ${name}`
});

router.of("message:received").use(context => {
    const { account } = context;
    const name = account.firstName ?? account.username
    context.reply(context.t.mymodule.hello(name));
});
```

## Плюрализация

Плюрализация достигается использованием функций-хелперов для каждого конкретного языка.
В `@replikit/i18n` существует встроенные хелперы для некоторых языков:

```ts
export interface Locale {
    mymodule: {
        apples(count: number): string;
    }
}

import { ru } from "@replikit/i18n";

locales.add("ru", "mymodule", {
    apples: count => ru.plural(count, "1 яблоко", "$ яблока", `$ яблок`)
});
```
