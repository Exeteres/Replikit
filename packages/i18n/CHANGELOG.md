# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.15.1](https://github.com/Exeteres/Replikit/compare/v0.15.0...v0.15.1) (2020-08-13)

**Note:** Version bump only for package @replikit/i18n





# [0.15.0](https://github.com/Exeteres/Replikit/compare/v0.14.0...v0.15.0) (2020-08-13)

**Note:** Version bump only for package @replikit/i18n





# [0.14.0](https://github.com/Exeteres/Replikit/compare/v0.13.0...v0.14.0) (2020-08-05)


### Bug Fixes

* **i18n:** Allow getLocale method for inline query contexts ([48c5836](https://github.com/Exeteres/Replikit/commit/48c5836d55294f777a0c6bb8ff605bd57e23e877))





# [0.13.0](https://github.com/Exeteres/Replikit/compare/v0.12.1...v0.13.0) (2020-08-04)


### Bug Fixes

* **i18n:** Fix bug with invalid locale resolution in MemberContext ([a1af23d](https://github.com/Exeteres/Replikit/commit/a1af23d764f417dcbceef3eba877d28b6e0b8b03))





# [0.12.0](https://github.com/Exeteres/Replikit/compare/v0.11.2...v0.12.0) (2020-07-30)

**Note:** Version bump only for package @replikit/i18n





## [0.11.2](https://github.com/Exeteres/Replikit/compare/v0.11.1...v0.11.2) (2020-07-24)

**Note:** Version bump only for package @replikit/i18n





## [0.11.1](https://github.com/Exeteres/Replikit/compare/v0.11.0...v0.11.1) (2020-07-24)

**Note:** Version bump only for package @replikit/i18n






# [0.11.0](https://github.com/Exeteres/Replikit/compare/v0.10.1...v0.11.0) (2020-07-24)


### Features

* **i18n:** Replace way to use i18n ([df6abcd](https://github.com/Exeteres/Replikit/commit/df6abcd791644bc2255107d98497a7e492fb07ca))





## [0.10.1](https://github.com/Exeteres/Replikit/compare/v0.10.0...v0.10.1) (2020-06-30)

**Note:** Version bump only for package @replikit/i18n






# [0.10.0](https://github.com/Exeteres/Replikit/compare/v0.9.0...v0.10.0) (2020-06-29)

**Note:** Version bump only for package @replikit/i18n





# [0.9.0](https://github.com/Exeteres/Replikit/compare/v0.8.0...v0.9.0) (2020-06-28)

**Note:** Version bump only for package @replikit/i18n





# [0.8.0](https://github.com/Exeteres/Replikit/compare/v0.7.2...v0.8.0) (2020-06-27)

**Note:** Version bump only for package @replikit/i18n





## [0.7.2](https://github.com/Exeteres/Replikit/compare/v0.7.1...v0.7.2) (2020-06-23)

**Note:** Version bump only for package @replikit/i18n





# [0.7.0](https://github.com/Exeteres/Replikit/compare/v0.6.0...v0.7.0) (2020-06-23)

**Note:** Version bump only for package @replikit/i18n





# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)

**Note:** Version bump only for package @replikit/i18n





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/i18n





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/i18n





# 0.5.0 (2020-06-20)


### Bug Fixes

* **i18n:** Fix invalid ru pluralization ([1ee3de9](https://github.com/Exeteres/Replikit/commit/1ee3de99923b1321ceb87af0eb8c1a3cb7884df5))


### Features

* **cli:** Add build command ([5c9dc57](https://github.com/Exeteres/Replikit/commit/5c9dc57b9ac2428f4970096dcfc6ae75b2e4dd27))






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)


### Features

* **cli:** Add build command ([0212d7f](https://github.com/Exeteres/Replikit/commit/0212d7f9c87d517b8cf220be9b32376551c3603b))





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)


### Bug Fixes

* **i18n:** Fix invalid ru pluralization ([fcb1599](https://github.com/Exeteres/Replikit/commit/fcb15995e152b7b901838dc468eff5f02f2cf400))





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)

**Note:** Version bump only for package @replikit/i18n
