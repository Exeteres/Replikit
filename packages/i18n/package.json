{
    "name": "@replikit/i18n",
    "version": "0.11.0",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "tsc -p tsconfig.build.json"
    },
    "dependencies": {
        "@replikit/core": "^0.11.0",
        "@replikit/router": "^0.11.0"
    }
}
