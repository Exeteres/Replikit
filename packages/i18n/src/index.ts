/// <reference path="../typings/index.d.ts" />

export * from "./utils";
export * from "./errors";
export * from "./localeStorage";
export * from "./helpers/ru";

export * from "./extensions/accountContextExtension";
export * from "./extensions/channelContextExtension";

export * from "./startup";
