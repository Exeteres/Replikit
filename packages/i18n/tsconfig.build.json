{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/i18n": ["i18n/src"]
        }
    },
    "include": ["src"]
}
