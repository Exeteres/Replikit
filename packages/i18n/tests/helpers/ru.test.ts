import { ru } from "@replikit/i18n";

describe("ru helpers", () => {
    it("should pluralize string", () => {
        const results: string[] = [];
        for (let i = 1; i <= 30; i++) {
            const result = ru.plural(i, "$ яблоко", "$ яблока", "$ яблок");
            results.push(result);
        }
        expect(results).toMatchSnapshot();
    });
});
