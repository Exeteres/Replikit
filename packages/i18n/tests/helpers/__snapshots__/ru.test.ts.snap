// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ru helpers should pluralize string 1`] = `
Array [
  "-30 яблок",
  "-29 яблок",
  "-28 яблок",
  "-27 яблок",
  "-26 яблок",
  "-25 яблок",
  "-24 яблока",
  "-23 яблока",
  "-22 яблока",
  "-21 яблоко",
  "-20 яблок",
  "-19 яблок",
  "-18 яблок",
  "-17 яблок",
  "-16 яблок",
  "-15 яблок",
  "-14 яблок",
  "-13 яблок",
  "-12 яблок",
  "-11 яблок",
  "-10 яблок",
  "-9 яблок",
  "-8 яблок",
  "-7 яблок",
  "-6 яблок",
  "-5 яблок",
  "-4 яблока",
  "-3 яблока",
  "-2 яблока",
  "-1 яблоко",
  "0 яблок",
  "1 яблоко",
  "2 яблока",
  "3 яблока",
  "4 яблока",
  "5 яблок",
  "6 яблок",
  "7 яблок",
  "8 яблок",
  "9 яблок",
  "10 яблок",
  "11 яблок",
  "12 яблок",
  "13 яблок",
  "14 яблок",
  "15 яблок",
  "16 яблок",
  "17 яблок",
  "18 яблок",
  "19 яблок",
  "20 яблок",
  "21 яблоко",
  "22 яблока",
  "23 яблока",
  "24 яблока",
  "25 яблок",
  "26 яблок",
  "27 яблок",
  "28 яблок",
  "29 яблок",
  "30 яблок",
]
`;
